File: src/main/java/com/kasisoft/libs/common/functional/KRunnable.java
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import jakarta.validation.constraints.*;

/**
 * @author daniel.kasmeroglu@kasisoft.com
 */
@FunctionalInterface
public interface KRunnable {

    void run() throws Exception;

    @NotNull
    default Runnable protect() {
        return () -> {
            try {
                run();
            } catch (Exception ex) {
                throw KclException.wrap(ex);
            }
        };
    }

} /* ENDINTERFACE */

File: src/main/kotlin/com/kasisoft/libs/common/functional/KRunnable.kt
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

/**
 * This implementation allows to throw Exceptions (useful on the Java side)
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
fun interface KRunnable {

    @Throws(Exception::class)
    fun run()

    fun protect(): Runnable =
        Runnable {
            try {
                run()
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

} /* ENDINTERFACE */
